subsys/bluetooth/audio/tbs.c

```diff
@@ -0,0 +1,2481 @@
+/* Bluetooth TBS - Telephone Bearer Service
+ *
+ * Copyright (c) 2020 Bose Corporation
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <zephyr/types.h>
+
+#include <device.h>
+#include <init.h>
+#include <stdlib.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+
+#include "tbs_internal.h"
+#include "ccid_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_TBS)
+#define LOG_MODULE_NAME bt_tbs
+#include "common/log.h"
+
+#define BT_TBS_VALID_STATUS_FLAGS(val)         ((val) <= (BIT(0) | BIT(1)))
+
+struct tbs_service_inst {
+	/* Attribute values */
+	char provider_name[CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH];
+	char uci[BT_TBS_MAX_UCI_SIZE];
+	char uri_scheme_list[CONFIG_BT_TBS_MAX_SCHEME_LIST_LENGTH];
+	uint8_t technology;
+	uint8_t signal_strength;
+	uint8_t signal_strength_interval;
+	uint8_t ccid;
+	uint16_t optional_opcodes;
+	uint16_t status_flags;
+	struct bt_tbs_in_uri incoming_uri;
+	struct bt_tbs_terminate_reason terminate_reason;
+	struct bt_tbs_in_uri friendly_name;
+	struct bt_tbs_in_uri in_call;
+
+	/* Instance values */
+	uint8_t index;
+	struct bt_tbs_call calls[CONFIG_BT_TBS_MAX_CALLS];
+	bool notify_current_calls;
+	bool notify_call_states;
+	bool pending_signal_strength_notification;
+	struct k_work_delayable reporting_interval_work;
+
+	/* TODO: The TBS (service) and the Telephone Bearers should be separated
+	 * into two different instances. This is due to the addition of GTBS,
+	 * where we now are in a state where this isn't a 1-to-1 correlation
+	 * between TBS and the Telephone Bearers
+	 */
+	struct bt_gatt_service *service_p;
+};
+
+struct gtbs_service_inst {
+	/* Attribute values */
+	char provider_name[CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH];
+	char uci[BT_TBS_MAX_UCI_SIZE];
+	uint8_t technology;
+	uint8_t signal_strength;
+	uint8_t signal_strength_interval;
+	uint8_t ccid;
+	uint16_t optional_opcodes;
+	uint16_t status_flags;
+	struct bt_tbs_in_uri incoming_uri;
+	struct bt_tbs_in_uri friendly_name;
+	struct bt_tbs_in_uri in_call;
+
+	/* Instance values */
+	bool notify_current_calls;
+	bool notify_call_states;
+	bool pending_signal_strength_notification;
+	struct k_work_delayable reporting_interval_work;
+
+	/* TODO: The TBS (service) and the Telephone Bearers should be separated
+	 * into two different instances. This is due to the addition of GTBS,
+	 * where we now are in a state where this isn't a 1-to-1 correlation
+	 * between TBS and the Telephone Bearers
+	 */
+	const struct bt_gatt_service_static *service_p;
+};
```

Comment:
It seems to me like it might make sense to make the GTBS struct a parent struct of the TBS struct, i.e. so that you have a GTBS struct as the first element of the TBS struct and the TBS struct that way extends the GTBS struct. This would help eliminate the void pointer stuff you have with casting between these types (you'd always pass a GTBS pointer and then when TBS is needed you use CONTAINER_OF() to get it).

Reply:
Yes, there certainly are a lot of common fields between the two. 

I tried doing a common struct before, but the number of differences between the two made me split it into two, but maybe I should revisit the idea. 
List of differences:

- uri_scheme_list only in TBS
- terminate_reason only in TBS
- index only in TBS
- calls only in TBS

So it would definitely make sense to have a base struct for GTBS and have the TBS struct use that. 

I'll add a note at the struct for this.